### PR TITLE
Conditional build

### PR DESCRIPTION
**Pull request by PorterNan**
# What

Merge conditional build into @prprabhu-ms's branch

- Add overall conditional build for communication-react package
- Add build:public and test:public command for each sub-project
- Add global bulk command - rush build:public and rush test:public
- Change comment block style to one line annotation comment

**Also put 2 updated api.md files as the example of how it looks after conditional build, free feel to remove/revert them**

TODO:
- Add conditional build into UI Tests (will be another PR)
- Add conditional build into pipeline (another PR)

# Why

# How Tested

# Process & policy checklist

- [ ] I have updated the project documentation to reflect my changes if necessary.
- [ ] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [ ] This change causes current functionality to break.

### CODE REVIEW

File: common/config/rush/command-line.json [hunk 1]
```diff
@@ -31,6 +31,25 @@
       "shellCommand": "node common/scripts/ensure-consistent-versions.js",
       "safeForSimultaneousRushProcesses": true
     },
+    {
+      "commandKind": "bulk",
+      "name": "build:public",
+      "summary": "Preprocess and build the non-beta version.",
```
Comment (prprabhu-ms) [inline common/config/rush/command-line.json:37]: nitty nit: Reword to be positive (what's non-beta, does it mean nightly?): "Build the @public variant of all packages"

Let's also consistently choose a word to refer to this: `version` vs `variant` vs `flavour`.
I'd argue against `version` (version refers to SemVer). I prefer `variant` over `flavour` because I don't want to get into endless fights about British vs American spelling :P

Comment (prprabhu-ms) [reply common/config/rush/command-line.json:37]: Along those lines, could we also rename this command to `build-variant:public` and `test-variant:public`? By itself, `build:public` is a bit confusing (it makes me think there's a private thing I could build instead).

File: common/scripts/babel-conditional-preprocess.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
+Object.defineProperty(exports, '__esModule', { value: true });
```
Comment (prprabhu-ms) [inline common/scripts/babel-conditional-preprocess.js:1]: Please add a top-level comment explaining what this preprocessing step does, and where it is used.

**Comment by prprabhu-ms** [comment]
Partial review. Let's walk through the Babel plugin together.

We should add a doc explaining how the plugin works, and how to do beta vs. public development

File: common/config/rush/command-line.json (hunk 1 above)
Comment (PorterNan) [reply common/config/rush/command-line.json:37]: For build:public one, maybe we should call it build:prod? I do think build:public sounds like there is a private version, and our beta is still "public", so maybe prod or ga would be something more accurate?

Comment (PorterNan) [reply common/config/rush/command-line.json:37]: I come up with another term build:stable that sounds like a good term for GA, so we might have alpha, beta then stable for GA

File: common/scripts/babel-conditional-preprocess.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+const helper = require('@babel/helper-plugin-utils');
+const t = require('@babel/types');
+exports.default = helper.declare((_api, opts) => {
+  const {
+    annotations
+  } = opts;
+
+  function Handle(path) {
+    let { node } = path;
```
Comment (prprabhu-ms) [inline common/scripts/babel-conditional-preprocess.js:11]: Please add link to AST types: https://babeljs.io/docs/en/babel-types#typescript

File: common/scripts/babel-conditional-preprocess.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    let removed = false;
+
+    for (const annotation of annotations) {
+      const {
+        match
+      } = annotation;
+
+      if (node.leadingComments && node.leadingComments.length > 0) {
```
Comment (prprabhu-ms) [inline common/scripts/babel-conditional-preprocess.js:19]: If possible, link to documentation for Node type that shows `leadingComment` always exists.

File: common/scripts/babel-conditional-preprocess.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        for (const comment of node.leadingComments) {
+          if (comment.value.includes(match)) {
```
Comment (prprabhu-ms) [inline common/scripts/babel-conditional-preprocess.js:21]: Single `if`?

File: common/scripts/babel-conditional-preprocess.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            removed = true;
+          }
+          if(removed) {
+            comment.ignore = true;
+          }
+        }
+      }
+
+      if (removed) {
```
Comment (prprabhu-ms) [inline common/scripts/babel-conditional-preprocess.js:30]: Either:

* Add comments explaining what is happening here ++ only remove if `path.parentPath` has only one child.

Or:

* Just document that we don't support JSX expressions. We'd have to create private Component instead.

File: common/scripts/babel-conditional-preprocess.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
+Object.defineProperty(exports, '__esModule', { value: true });
+
+const helper = require('@babel/helper-plugin-utils');
+const t = require('@babel/types');
```
[the remark this answers is not in the document]
Comment (PorterNan) [reply common/scripts/babel-conditional-preprocess.js:4]: Agree with babelHelper, but maybe will keep this one. This 't' is like a recommendation name from babel plugin handbook - https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/plugin-handbook.md

Which is also widely used by other plugin developers

File: common/scripts/babel-conditional-preprocess.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+      ExportAllDeclaration(path) {
+        Handle(path);
+      },
+
+      TSPropertySignature(path) {
+        Handle(path);
+      },
+
+      Expression(path) {
```
[the remark this answers is not in the document]
Comment (PorterNan) [reply common/scripts/babel-conditional-preprocess.js:75]: Sync offline, we still need it to make some special cases work, but we need to avoid using it a lot

File: common/scripts/babel-conditional-preprocess.js [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+      FunctionDeclaration(path) {
+        Handle(path);
+      },
+
+      Statement(path) {
+        Handle(path);
+      },
+
+      VariableDeclaration(path) {
```
[the remark this answers is not in the document]
Comment (PorterNan) [reply common/scripts/babel-conditional-preprocess.js:55]: Sync offline, will keep this to pass vscode 'used' compiler check in some cases

File: common/scripts/babel-conditional-preprocess.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
+Object.defineProperty(exports, '__esModule', { value: true });
+
+const helper = require('@babel/helper-plugin-utils');
+const t = require('@babel/types');
+exports.default = helper.declare((_api, opts) => {
+  const {
+    annotations
```
[the remark this answers is not in the document]
Comment (PorterNan) [reply common/scripts/babel-conditional-preprocess.js:10]: I was about to apply this comment, but after thinking for a while I'd say I would prefer to keep it, the main reason is people don't have to read conditional-complie.js to see/modify this magic keyword, the keyword should be able to be found within the project (this creates a direct coupling to share a string between project and plugin)- for avoiding setting the wrong annotation, I will use a common config file to put it just for once

File: common/scripts/babel-conditional-preprocess.js (hunk 5 above)
Comment (PorterNan) [reply common/scripts/babel-conditional-preprocess.js]: I found a logic leak if we use if - we wanna remove all the rest following comments after the annotation comment, which means removed = true but value doesn't match, reverting this back

File: packages/acs-ui-common/package.json [hunk 11]
```diff
@@ -11,6 +11,12 @@
     "build:cjs": "rollup -c --silent --failAfterWarnings",
     "build:watch": "rushx build",
     "clean": "rimraf dist",
+    "copy-original": "cpx ./src/** ./preprocessed",
```
Comment (prprabhu-ms) [inline packages/acs-ui-common/package.json:14]: Why would we want to run `rushx copy-original`, ever?

I'd drop this command

File: packages/acs-ui-common/package.json [hunk 12]
```diff
@@ -11,6 +11,12 @@
...
+    "preprocess": "rushx copy-original && babel ./src --out-dir ../preprocessed --extensions \".ts,.tsx\" --keep-file-extension --config-file ./.babelrc.js --relative && rimraf ../preprocessed",
```
Comment (prprabhu-ms) [inline packages/acs-ui-common/package.json:15]: Do we ever want to run `rushx preprocess` instead of `rushx preprocess:stable`? I'd drop this one as well.

File: packages/react-components/src/components/MessageThread.tsx [hunk 13]
```diff
@@ -261,7 +261,7 @@ const DefaultSystemMessage: MessageRenderer = (props: MessageProps) => {
         case 'content':
           return (
             <SystemMessageComponent
-              iconName={(message.iconName ?? '') as SystemMessageIconTypes}
+              iconName={(message.iconName ? message.iconName : '') as SystemMessageIconTypes}
```
Comment (prprabhu-ms) [inline packages/react-components/src/components/MessageThread.tsx:264]: Please remove this unrelated change.

File: samples/Calling/package.json [hunk 14]
```diff
@@ -5,12 +5,14 @@
   "description": "This is a sample application to show how the UI Library can be used to build a calling experience. The client-side application is a React based user interface which uses React Hooks for handling complex state while leveraging Microsoft Fluent UI for how things look.",
   "scripts": {
     "build": "rushx clean && concurrently --kill-others-on-fail \"tsc\" \"webpack --mode=production --env production\"",
+    "build:stable": "echo skip",
```
Comment (prprabhu-ms) [inline samples/Calling/package.json:8]: In a follow up PR: We need a way to  build these for both falvors as well.

File: packages/storybook/package.json [hunk 15]
```diff
@@ -4,10 +4,12 @@
   "description": "Azure Communication Services UI Library Storybook",
   "scripts": {
     "build": "build-storybook -s stories,public,.storybook  --quiet --loglevel warn",
+    "build:stable": "echo skip",
```
Comment (prprabhu-ms) [inline packages/storybook/package.json:7]: Follow up PR: We need a way to build / start these for both flavors as well.

**Review by prprabhu-ms — APPROVED** [review]
Let's merge this and follow up about:
* Storybook: Build, test, start for both flavors
* Samples: Build, test, start for both flavors
* e2e tests: Build, run for both flavors
